### PR TITLE
feat(aws-codepipeline): support for pipeline action’s service role

### PR DESCRIPTION
**Pull request by rsmogura**
The action’s service roles is a role which will be assumed
by pipeline during execution of this action.

The pipeline action’s service role can be used to perform more
advanced configuration, when i.e. elevation of permissions
is required, or when fine grained access control may be required.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codepipeline-pipeline-stages-actions.html

This commit is motivated by enabling cross-account deployments,
for which service role’s will be used as jump role to assume
one used by Cloud Formation in target account.

BREAKING CHANGE: `PipelineCloudFormationDeployActionProps.role`,
`PipelineCloudFormationDeployAction.role`,
`PipelineDeployStackActionProps.role` renamed to `deploymentRole`
----

### Pull Request Checklist

* [X ] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [ X] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [X ] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [ ] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts [hunk 1]
```diff
@@ -151,6 +151,15 @@ export interface CommonActionConstructProps {
...
   stage: IStage;
+
+  /**
+   * The service role that is assumed during execution of action.
+   * This role is not mandatory, however more advanced configuration
+   * may require specifying it.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codepipeline-pipeline-stages-actions.html
+   */
+  actionRole?: iam.IRole;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: Can we just call this `role`?

Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: `PipelineCloudFormationDeployActionProps` uses `role` attribute to describe role assumed by Cloud Formation. The ref. arch. actually suggest usage of both roles for cross-account deployments.

If we would rename this attribute, one on *CloudFormation* should be renamed too, which would occur in a backward incompatible change.

**Comment by skinny85** [comment]
Thanks for the PR @rsmogura! The one thing I would change is to add this new `actionRole` property to the `CommonActionProps` interface defined in `aws-codepipeline-api` package. This way, it will be available to all Actions, not only to the CloudFormation ones.

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: I wonder whether it's actually worth it to change the attribute in `PipelineCloudFormationDeployActionProps` to something like `cloudFormationRole`, and name this new one `role`. The reason is that this one is much more generic than the CFN Actions one, so the name `role` there does not seem entirely appropriate.

@eladb what do you think?

Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: @skinny85 not a bad idea. Maybe something like `deploymentRole`? (it is a deploy action)

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts [hunk 2]
```diff
@@ -141,6 +141,15 @@ export interface CommonActionProps {
...
   runOrder?: number;
+
+  /**
+   * The service role that is assumed during execution of action.
+   * This role is not mandatory, however more advanced configuration
+   * may require specifying it.
+   *
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codepipeline-pipeline-stages-actions.html
+   */
+  actionRole?: iam.IRole;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: This is a good start :) but is not enough to be passed from every role back to the `Role` class. You have to make sure the constructors of every concrete Action class call `super` with the correct parameters. The best way is to do it is to use the [spread operator](https://basarat.gitbooks.io/typescript/docs/spread-operator.html), like this:

```ts
class SomeAction extends AnotherAction {
  constructor(parent: cdk.Construct, id: string, props: SomeActionProps) {
    super(parent, id, {
      // some properties passed explicitly
      ...props,
    });
  }
}
```

This way, you're guaranteed all of the properties will be automatically passed (the reason some of them don't use it yet is because it was written for an older TypeScript version).

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: `deploymentRole` sounds a little ambiguous to me, but sure. @rsmogura can you make this change? Thanks!

Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: I think the name like `deploymentRole` would look ok (refarch for cross account deployments uses similar term *deployer*), and definitely it should be changed.

I would like to point out that if we will call the `actionRole` a `role` it can be backward incompatible and hardly captured by TypeScript transpiler - if someone in existing code already have set a `role` for CF action after update behavior of this property will change. So I just would like to be sure if it's ok.

I think we should as well consider changing and moving `addToRolePolicy` to `Action`, and changing this method on CF action to something like `addToDeploymentRolePolicy`?

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 2 above)
Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: I added passing the action's role for other actions. I wonder if we would like to refactor existing code to use spread operator, so in future adding new attributes to action's interfaces would be more transparent?

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 1 above)
Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: I renamed CF `role` to `deploymentRole` and `PipelineDeployStackAction` in `app-delivery`, changed `addToRolePolicy` method names to `addToDeploymentRolePolicy`.

I do not update action's service role permissions when added to pipeline. I think as feature is for advanced use-cases role or artifacts can come from other stack(s).

I tested this change with simple pipeline and setting action's role on source action, requires adding **s3** and **codecommit** actions to be granted by developer.

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: Yes, every Action subclass that does not currently use the spread operator should be changed to use it. I thought I made it clear in my previous comment, but I guess I failed :(

Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: Replaced
```
stage: props.stage,
runOrder: props.runOrder
```
with
```
...props
```

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts [hunk 3]
```diff
@@ -136,6 +136,15 @@ export interface CommonActionProps {
...
```
[the remark this answers is not in the document]
Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: Done. Please double check.

**Comment by RomainMuller** [comment]
@skinny85 - no major objection with the rename. I wonder if `deployRole` wouldn't be slightly better, but either is fine.

File: packages/@aws-cdk/aws-codepipeline-api/lib/action.ts (hunk 1 above)
Comment (RomainMuller) [reply packages/@aws-cdk/aws-codepipeline-api/lib/action.ts]: @skinny85  - I don't think of `deploymentRole` as so ambiguous in that context... Thats the way at least I think of it there... And it maps nicely with the user intention... So that's good. As I mentioned in a separate note, `deployRole` might be another option (I have no strong feelings either way).

File: packages/@aws-cdk/aws-codepipeline/test/test.pipeline.ts [hunk 4]
```diff
@@ -518,7 +531,7 @@ export = {
       test.done();
     },
   },
-};
+} as any;
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-codepipeline/test/test.pipeline.ts]: I'm suspicious why this one is necessary. I'd remove it as it "disables" type checking in ways that can hide other issues (see below).

Last I saw this done, it was motivated by the TypeScript compiler being confused by a genuine coding error somewhere else (it was a use of `this` - instead of some `testStack` - as the first argument of a `Construct` class instantiation caused `tsc` to assume the whole test group to be a `Construct`- which is obviously not the case - and that caused a very confusing error message to be issued).

Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline/test/test.pipeline.ts]: I will send update PR. I don't remember this situation exactly, TS was complaining that test attribute `'test name'` can't be is not some other class (maybe `Construct`). 

It was very wired. I think it was due to import of some other files, which I removed after clean build (I think some circular dependencies).

**Comment by rsmogura** [comment]
I've found that project build action is using `role` to represent build role - I think this should be refactored too.

**Comment by rsmogura** [comment]
> > I've found that project build action is using role to represent build role - I think this should be refactored too.
> 
> Can you link to the code?

My bad - I took a look at `Project` which doesn't inherit from action. Please ignore this remark.

File: packages/@aws-cdk/aws-lambda/lib/pipeline-action.ts [hunk 5]
```diff
@@ -84,6 +84,7 @@ export class PipelineInvokeAction extends codepipeline.Action {
   constructor(scope: cdk.Construct, id: string, props: PipelineInvokeActionProps) {
     super(scope, id, {
       stage: props.stage,
+      role: props.role,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-lambda/lib/pipeline-action.ts]: Use the spread operator here as well.

File: packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json [hunk 6]
```diff
@@ -221,7 +215,13 @@
...
             "Name": "CFN"
           }
-        ]
+        ],
+        "ArtifactStore": {
+          "Location": {
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json]: Can you revert this change? Ordering should not matter for integ tests.

Comment (rsmogura) [reply packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json]: Done

**Comment by rix0rrr** [comment]
What do we do with this PR? @skinny85 ?

**Comment by skinny85** [comment]
@rix0rrr yep, I'm on this one, no worries

File: packages/@aws-cdk/aws-codecommit/test/test.codecommit.ts [hunk 7]
```diff
@@ -49,7 +49,7 @@ export = {
       test.throws(() => myRepository.notify('myTrigger'));
 
       test.done();
-    }
+    },
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codecommit/test/test.codecommit.ts]: I think you can safely revert changes to this file :)

File: packages/@aws-cdk/aws-codedeploy/test/test.pipeline-action.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+class TestApp {
+  private readonly app = new App();
+  // tslint:disable-next-line:member-ordering
+  public readonly stack: Stack = new Stack(this.app, 'MyStack');
+
+  public synthesizeTemplate() {
+    return this.app.synthesizeStack(this.stack.name).template;
+  }
+}
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codedeploy/test/test.pipeline-action.ts]: Is this new test needed?

File: packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts [hunk 9]
```diff
@@ -62,7 +62,8 @@ export = {
     stage: prodStage,
     stackName,
     changeSetName,
-    role: changeSetExecRole,
+    runOrder: 321,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts]: Super minor, but why add this `runOrder` here? Seems you can safely test this in a different place, like the new test you added below.

File: packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts [hunk 10]
```diff
@@ -356,6 +357,53 @@ export = {
...
+        }, {
+          "Name": "FreshRoleAction",
+          "RoleArn": {
+            "Fn::GetAtt": [
+              "FreshRole472F6E18",
+              "Arn"
+            ]
+          }
+        }]
+      }]
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/test.cloudformation-pipeline-actions.ts]: Can you use standard JS indentation here instead? I have a hard time following what's supposed to be asserted here.

```ts
    expect(stack).to(haveResourceLike('AWS::CodePipeline::Pipeline', {
      "Stages": [
        {
          "Name": "Source" /* don't care about the rest */
        },
        {
          "Name": "Deploy",
          "Actions": [
            {
              "Name": "ImportedRoleAction",
              "RoleArn": "arn:aws:iam::000000000000:role/action-role"
            },
            {
              "Name": "FreshRoleAction",
              "RoleArn": {
                "Fn::GetAtt": [
                  "FreshRole472F6E18",
                  "Arn"
                ],
              }
            }
          ]
        }
      ]
    }));
```

File: packages/@aws-cdk/aws-codepipeline/test/test.pipeline.ts [hunk 11]
```diff
@@ -102,7 +102,7 @@ export = {
             "Name": "A"
           }
           ],
-          "RunOrder": 8
+          "RunOrder": 8,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codepipeline/test/test.pipeline.ts]: Remove this change :D

**Comment by rsmogura** [comment]
I will try to sit on it today or tomorrow, unfortunately today I had a lot of project work.

File: packages/@aws-cdk/aws-codedeploy/test/test.pipeline-action.ts (hunk 8 above)
Comment (rsmogura) [reply packages/@aws-cdk/aws-codedeploy/test/test.pipeline-action.ts]: I think I used it when `role` was on Action props. Haven't spotted it's mine addition. Removing this file.

**Comment by skinny85** [comment]
Thanks @rsmogura! Merged.
